### PR TITLE
Batched delegations

### PR DESCRIPTION
**Pull request by sampocs**
## Context
In order to support validator sets of 100+, we have to batch delegations so that we don't exceed the gas limit. This PR handles the changes in the delegation submission. The callback will be in a separate PR.

**Note to reviewer**: I'd recommend reviewing commit by commit

## Brief Changelog
* Added `DelegationTxsInProgress` field to deposit record for retry protection - it's incremented when delegation is submitted and will be decremented in the callback
* Split `DelegateOnHost` into two functions: `GetDelegationICAMessages` and `BatchSubmitDelegationICAs`. The former builds up all the batches across validators, and the latter submits the ICAs in batches
* Removed the max cap on deposit records - it's unnecessary 
* Added unit tests

### CODE REVIEW

**Comment by sampocs** [comment]
@ethan-stride re: combining deposit records

I think during the design phase we landed on sending messages per one record for simplicity (.e.g. no loops) - I don't think submitting an extra ICA to each validator is too concerning. 

Long term, what I think we should really do is combine the two deposit records into 1 (e.g. instead of creating a new record from rewards, we just add the amount to the existing record). And then we can keep the code as is, but it essentially accomplishes the same goal your after. But I think this is outta scope at this point

File: x/stakeibc/keeper/delegation.go [hunk 1]
```diff
@@ -101,39 +100,23 @@ func (k Keeper) StakeExistingDepositsOnHostZones(ctx sdk.Context, epochNumber ui
...
-		if hostZone.Halted {
-			k.Logger(ctx).Error(fmt.Sprintf("[StakeExistingDepositsOnHostZones] Host zone halted for deposit record {%d}", depositRecord.Id))
-			continue
-		}
-
-		if hostZone.DelegationIcaAddress == "" {
-			k.Logger(ctx).Error(fmt.Sprintf("[StakeExistingDepositsOnHostZones] Zone %s is missing a delegation address!", hostZone.ChainId))
+		hostZone, err := k.GetActiveHostZone(ctx, depositRecord.HostZoneId)
+		if err != nil {
+			k.Logger(ctx).Error(fmt.Sprintf("[StakeExistingDepositsOnHostZones] Not processing %d, %s", depositRecord.Id, err.Error()))
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go:132]: we sure did haha. I wasn't sure whether to remove it now or later

File: x/stakeibc/keeper/delegation.go [hunk 2]
```diff
@@ -14,21 +14,25 @@ import (
 	"github.com/Stride-Labs/stride/v22/x/stakeibc/types"
 )
 
-func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk.Coin, depositRecord recordstypes.DepositRecord) error {
+// Builds the delegation ICA messags for a given deposit record
+// Each validator has a portion of the total amount on the record based on their weight
+func (k Keeper) GetDelegationICAMessages(
+	ctx sdk.Context,
+	hostZone types.HostZone,
+	amt sdk.Coin, depositRecord recordstypes.DepositRecord,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go]: i think this was resolved in a later commit

File: x/stakeibc/keeper/delegation.go [hunk 3]
```diff
@@ -14,21 +14,25 @@ import (
...
+) (msgs []proto.Message, delegations []*types.SplitDelegation, err error) {
 	// Fetch the relevant ICA
 	if hostZone.DelegationIcaAddress == "" {
-		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", hostZone.ChainId)
+		return msgs, delegations, errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", hostZone.ChainId)
 	}
 
 	// Construct the transaction
 	targetDelegatedAmts, err := k.GetTargetValAmtsForHostZone(ctx, hostZone, amt.Amount)
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go]: yeah it was in host_zone cause it's shared across the two files and can more generally be thought of as a function that takes an input amount and splits it up into the relevant portions for each validator based on weights. I also think we should rename it

File: x/stakeibc/keeper/delegation.go [hunk 4]
```diff
@@ -14,21 +14,25 @@ import (
...
 	if err != nil {
 		k.Logger(ctx).Error(fmt.Sprintf("Error getting target delegation amounts for host zone %s", hostZone.ChainId))
-		return err
+		return msgs, delegations, err
 	}
 
-	var splitDelegations []*types.SplitDelegation
-	var msgs []proto.Message
 	for _, validator := range hostZone.Validators {
 		relativeAmount, ok := targetDelegatedAmts[validator.Address]
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go]: I think this was done in a later commit

File: x/stakeibc/keeper/delegation.go [hunk 5]
```diff
@@ -40,40 +44,51 @@ func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk
...
+	depositRecord recordstypes.DepositRecord,
+	msgs []proto.Message,
+	delegations []*types.SplitDelegation,
+) (numTxsSubmitted uint64, err error) {
 	// add callback data
 	delegateCallback := types.DelegateCallback{
 		HostZoneId:       hostZone.ChainId,
 		DepositRecordId:  depositRecord.Id,
-		SplitDelegations: splitDelegations,
+		SplitDelegations: delegations,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go]: I always thought splitDelegations was kind of a dumb name, plus I was kind of following the same form as the unbonding code where I thought it was clear in the context of the function.

That said, not opposed to changing it back if you think it was more clear

File: x/stakeibc/keeper/delegation.go [hunk 6]
```diff
@@ -40,40 +44,51 @@ func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk
...
+func (k Keeper) BatchSubmitDelegationICAMessages(
+	ctx sdk.Context,
+	hostZone types.HostZone,
+	depositRecord recordstypes.DepositRecord,
+	msgs []proto.Message,
+	delegations []*types.SplitDelegation,
+) (numTxsSubmitted uint64, err error) {
 	// add callback data
 	delegateCallback := types.DelegateCallback{
 		HostZoneId:       hostZone.ChainId,
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go]: yeah good point, it's probably not necessary. Doesn't really hurt though imo

File: x/stakeibc/keeper/delegation_test.go [hunk 7]
```diff
@@ -403,3 +406,228 @@ func (s *KeeperTestSuite) TestStakeDepositRecords_NoDelegationAccount() {
...
+
+				// Check the callback
+				s.Require().Equal(expected.Validator, actualSplit.Validator, "callback validator for %s", valAddress)
+				s.Require().Equal(expected.Amount.Int64(), actualSplit.Amount.Int64(), "callback amount %s", valAddress)
+			}
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestBatchSubmitDelegationICAMessages() {
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/delegation_test.go:539]: I think we have these tests already in delegation_test.go for the test on `StakeExistingDepositRecords`

File: x/stakeibc/keeper/icacallbacks_delegate.go [hunk 8]
```diff
@@ -63,11 +62,21 @@ func (k Keeper) DelegateCallback(ctx sdk.Context, packet channeltypes.Packet, ac
 		return errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "deposit record not found %d", recordId)
 	}
 
-	// Regardless of failure/success/timeout, indicate that this ICA has completed
+	// Regardless of failure/success/timeout, indicate that this ICA has completed on the deposit record
+	if depositRecord.DelegationTxsInProgress == 0 {
+		return types.ErrInvalidDelegationsInProgress.Wrapf("delegation changes in progress is already 0 and can't be decremented")
+	}
+	depositRecord.DelegationTxsInProgress -= 1
```
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/icacallbacks_delegate.go:49]: that's only when you're dealing with an sdkmath.Int - it's a uint64 in this case

File: x/stakeibc/keeper/delegation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/stakeibc/keeper/delegation.go:132]: Your call if you want to remove it.  I think it makes things more consistent to remove it

File: x/stakeibc/keeper/delegation.go [hunk 9]
```diff
@@ -93,53 +111,56 @@ func (k Keeper) StakeExistingDepositsOnHostZones(ctx sdk.Context, epochNumber ui
...
+		return isStakeRecord && isBeforeCurrentEpoch && isNotInProgress
 	})
 
 	if len(stakeDepositRecords) == 0 {
 		k.Logger(ctx).Info("No deposit records in state DELEGATION_QUEUE")
 		return
 	}
 
-	// limit the number of staking deposits to process per epoch
-	maxDepositRecordsToStake := utils.Min(len(stakeDepositRecords), cast.ToInt(k.GetParam(ctx, types.KeyMaxStakeICACallsPerEpoch)))
```
Comment (riley-stride) [inline x/stakeibc/keeper/delegation.go:105]: Nit: should we remove the type?

File: x/stakeibc/keeper/delegation.go [hunk 10]
```diff
@@ -8,82 +8,100 @@ import (
...
 	"github.com/Stride-Labs/stride/v22/x/stakeibc/types"
 )
 
-func (k Keeper) DelegateOnHost(ctx sdk.Context, hostZone types.HostZone, amt sdk.Coin, depositRecord recordstypes.DepositRecord) error {
-	// Fetch the relevant ICA
-	if hostZone.DelegationIcaAddress == "" {
-		return errorsmod.Wrapf(types.ErrICAAccountNotFound, "no delegation account found for %s", hostZone.ChainId)
-	}
-
+// Builds the delegation ICA messags for a given deposit record
```
Comment (riley-stride) [inline x/stakeibc/keeper/delegation.go:17]: pico nit: messages

**Review by riley-stride — COMMENTED** [review]
Lgtm

File: x/stakeibc/keeper/delegation.go (hunk 10 above)
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go:17]: done

File: x/stakeibc/keeper/delegation.go (hunk 9 above)
Comment (sampocs) [reply x/stakeibc/keeper/delegation.go:105]: eh - there's a lot going on already so let's hold off for now. I think we can make it a separate ticket to cleanup our params cause I think we have a lot that aren't actually used

Comment (riley-stride) [reply x/stakeibc/keeper/delegation.go:105]: sgtm

**Review by shellvish — COMMENTED** [review]
Looks great, thanks a ton for doing this!

**Pull request merged**